File: ts-backend/src/rbac/rbac.ts
/**
 * Role-based access control (RBAC) for the TypeScript backend.
 *
 * Ports backend/rbac.py to TypeScript with exact behavioral parity.
 *
 * Tokens are stored as SHA-256 hashes in .autonomous-team/config.json under
 * the "rbac" key. Each token maps to a role name; each role defines an
 * allow-list of "METHOD /path-pattern" rules. Pattern matching uses the
 * fnmatch.ts port (SPIKE-2 proven: 0 divergences vs Python across 534 cases).
 *
 * LOAD-BEARING FALLTHROUGH (security-critical — must replicate exactly):
 *   - When rbac section is ABSENT from config: always allow (backward compat)
 *   - When rbac IS present but token hash NOT in key table: DENY (return false)
 *   - When rbac IS present but rbac section empty/null: allow-all mode
 *
 * This is the exact semantics of Python rbac.py check():
 *   if not self._enabled: return True       ← no rbac section → allow-all
 *   role_name = get_role_for_token(token)
 *   if role_name is None: return False      ← token not in key table → DENY
 *
 * Optional: memoize compiled regex per pattern for hot-path performance.
 * The fnmatch.ts fnmatchTranslate() compiles a RegExp — memoizing it avoids
 * recompilation on repeated RBAC checks for the same pattern.
 */

import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { fnmatch } from "./fnmatch.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface RoleDefinition {
  label?: string;
  allow: string[];
}

interface RbacConfig {
  roles?: Record<string, RoleDefinition>;
  keys?: Record<string, string>; // hash → role name
}

interface ConfigJson {
  rbac?: RbacConfig;
  [key: string]: unknown;
}

// ---------------------------------------------------------------------------
// Built-in roles — mirrors Python _BUILT_IN_ROLES in backend/rbac.py exactly
// ---------------------------------------------------------------------------
const BUILT_IN_ROLES: Record<string, RoleDefinition> = {
  admin: {
    label: "Administrator",
    allow: ["*"], // matches everything
  },
  agent: {
    label: "Agent (internal)",
    allow: [
      "GET /health",
      "GET /health/*",
      "GET /metrics",
      "GET /budget/*",
      "GET /registry",
      "GET /registry/*",
      "GET /agents",
      "GET /agents/*",
      "GET /kpi",
      "GET /kpi/*",
      "GET /stream/*",
      "GET /replays",
      "GET /replays/*",
      "GET /rbac/whoami",
      "POST /budget/init",
    ],
  },
  viewer: {
    label: "Read-only viewer",
    allow: [
      "GET *", // any GET is fine; POSTs are not listed
    ],
  },
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Return the hex-encoded SHA-256 digest of token. Mirrors Python _sha256(). */
export function sha256(token: string): string {
  return createHash("sha256").update(token, "utf8").digest("hex");
}

/**
 * Return True if method + path matches the allow-list rule.
 *
 * Rule forms (exact port of Python _match_rule()):
 *   "*"              — matches everything (any method, any path)
 *   "METHOD /glob"   — matches the given method and path glob
 *   "GET *"          — matches any GET regardless of path
 *
 * Path matching uses fnmatch (Python-equivalent semantics).
 */
export function matchRule(rule: string, method: string, path: string): boolean {
  rule = rule.trim();
  if (rule === "*") return true;

  const spaceIdx = rule.indexOf(" ");
  if (spaceIdx < 0) return false;

  const ruleMethod = rule.slice(0, spaceIdx);
  const rulePath = rule.slice(spaceIdx + 1);

  if (ruleMethod.toUpperCase() !== method.toUpperCase()) return false;
  return fnmatch(path, rulePath);
}

// ---------------------------------------------------------------------------
// RBACManager
// ---------------------------------------------------------------------------

export class RBACManager {
  private readonly configPath: string;
  private roles: Record<string, RoleDefinition> = {};
  private tokenHashes: Map<string, string> = new Map(); // hash → role name
  private _enabled = false;

  constructor(configPath: string) {
    this.configPath = configPath;
    this.load();
  }

  private load(): void {
    let raw: ConfigJson;
    try {
      const text = readFileSync(this.configPath, "utf8");
      raw = JSON.parse(text) as ConfigJson;
    } catch {
      return; // Missing or invalid config — backward-compatible allow-all mode
    }

    const rbac = raw.rbac;
    if (!rbac) {
      return; // No rbac section → backward-compatible allow-all mode
    }

    this._enabled = true;

    // Merge built-in roles with any overrides from config
    this.roles = { ...BUILT_IN_ROLES };
    if (rbac.roles) {
      for (const [roleName, roleDef] of Object.entries(rbac.roles)) {
        this.roles[roleName] = roleDef;
      }
    }

    // Index token hashes
    if (rbac.keys) {
      for (const [tokenHash, roleName] of Object.entries(rbac.keys)) {
        this.tokenHashes.set(tokenHash, roleName);
      }
    }
  }

  get enabled(): boolean {
    return this._enabled;
  }

  /** Return the role name for token, or null if unknown. */
  getRoleForToken(token: string): string | null {
    if (!this._enabled) return null;
    const h = sha256(token);
    return this.tokenHashes.get(h) ?? null;
  }

  /** Return the role definition dict, or null if unknown. */
  getRoleInfo(roleName: string): RoleDefinition | null {
    return this.roles[roleName] ?? null;
  }

  /**
   * Return true if token is allowed to call method on path.
   *
   * LOAD-BEARING FALLTHROUGH (must match Python rbac.py check() exactly):
   *   - When RBAC is disabled (no rbac section): always returns true
   *   - When RBAC is enabled AND token hash NOT in key table: returns false (DENY)
   *   - When role has a matching allow rule: returns true
   *   - Otherwise: returns false
   */
  check(token: string, method: string, path: string): boolean {
    if (!this._enabled) return true; // No rbac section → allow-all

    const roleName = this.getRoleForToken(token);
    if (roleName === null) return false; // Token not in key table → DENY

    const role = this.roles[roleName];
    if (!role) return false; // Unknown role → DENY

    for (const rule of role.allow) {
      if (matchRule(rule, method, path)) return true;
    }
    return false;
  }
}

File: backend/rbac.py
"""
Role-based access control (RBAC) for the fulcrumaxe REST API.

Tokens are stored as SHA-256 hashes in .autonomous-team/config.json under
the ``rbac`` key. Each token maps to a role name; each role defines an
allow-list of ``METHOD /path-pattern`` rules.  Pattern matching uses fnmatch
so ``*`` is a single-segment wildcard and ``**`` is not needed — just use
``GET /agents/*`` to allow any GET under /agents/.

Backward-compatible: if no ``rbac`` section exists in config, every
authenticated request is allowed.

Usage::

    manager = RBACManager(config_path)
    ok = manager.check(token, "GET", "/budget/status")
    if not ok:
        # return 403
"""

from __future__ import annotations

import fnmatch
import hashlib
import json
from pathlib import Path
from typing import Any


# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

_BUILT_IN_ROLES: dict[str, dict[str, Any]] = {
    "admin": {
        "label": "Administrator",
        "allow": ["*"],           # matches everything
    },
    "agent": {
        "label": "Agent (internal)",
        "allow": [
            "GET /health",
            "GET /health/*",
            "GET /metrics",
            "GET /budget/*",
            "GET /registry",
            "GET /registry/*",
            "GET /agents",
            "GET /agents/*",
            "GET /kpi",
            "GET /kpi/*",
            "GET /stream/*",
            "GET /replays",
            "GET /replays/*",
            "GET /rbac/whoami",
            "POST /budget/init",
        ],
    },
    "viewer": {
        "label": "Read-only viewer",
        "allow": [
            "GET *",              # any GET is fine; POSTs are not listed
        ],
    },
}


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _sha256(token: str) -> str:
    """Return the hex-encoded SHA-256 digest of *token*."""
    return hashlib.sha256(token.encode("utf-8")).hexdigest()


def _match_rule(rule: str, method: str, path: str) -> bool:
    """Return True if *method* + *path* matches the allow-list *rule*.

    Rule forms:
        ``*``              — matches everything (any method, any path)
        ``METHOD /glob``   — matches the given method and path glob
        ``GET *``          — matches any GET regardless of path

    Path matching uses :func:`fnmatch.fnmatch`.
    """
    rule = rule.strip()
    if rule == "*":
        return True
    parts = rule.split(" ", 1)
    if len(parts) != 2:
        return False
    rule_method, rule_path = parts
    if rule_method.upper() != method.upper():
        return False
    return fnmatch.fnmatch(path, rule_path)


# ---------------------------------------------------------------------------
# Public class
# ---------------------------------------------------------------------------


class RBACManager:
    """Load RBAC config and answer access-control queries.

    Parameters
    ----------
    config_path:
        Path to ``.autonomous-team/config.json``.
    """

    def __init__(self, config_path: Path | str) -> None:
        self._config_path = Path(config_path)
        self._roles: dict[str, dict[str, Any]] = {}
        self._token_hashes: dict[str, str] = {}   # hash → role name
        self._enabled: bool = False
        self._load()

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _load(self) -> None:
        """Parse config.json and populate internal tables."""
        try:
            raw = json.loads(self._config_path.read_text())
        except (OSError, json.JSONDecodeError):
            return

        rbac = raw.get("rbac")
        if not rbac:
            return   # no rbac section → backward-compatible allow-all mode

        self._enabled = True

        # Merge built-in roles with any overrides from config.
        merged_roles: dict[str, dict[str, Any]] = dict(_BUILT_IN_ROLES)
        for role_name, role_def in rbac.get("roles", {}).items():
            merged_roles[role_name] = role_def
        self._roles = merged_roles

        # Index token hashes.
        for token_hash, role_name in rbac.get("keys", {}).items():
            self._token_hashes[token_hash] = role_name

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    @property
    def enabled(self) -> bool:
        """True when an rbac section is present in config."""
        return self._enabled

    def get_role_for_token(self, token: str) -> str | None:
        """Return the role name for *token*, or None if unknown."""
        if not self._enabled:
            return None
        h = _sha256(token)
        return self._token_hashes.get(h)

    def get_role_info(self, role_name: str) -> dict[str, Any] | None:
        """Return the role definition dict, or None if the role is unknown."""
        return self._roles.get(role_name)

    def check(self, token: str, method: str, path: str) -> bool:
        """Return True if *token* is allowed to call *method* on *path*.

        When RBAC is disabled (no ``rbac`` section in config), always returns
        True so callers with a valid bearer token continue to work unchanged.

        Returns False when:
        - RBAC is enabled and the token hash is not in the key table.
        - The resolved role has no matching allow rule for method + path.
        """
        if not self._enabled:
            return True

        role_name = self.get_role_for_token(token)
        if role_name is None:
            return False

        role = self._roles.get(role_name)
        if role is None:
            return False

        for rule in role.get("allow", []):
            if _match_rule(rule, method, path):
                return True
        return False
